internal/security/common/usermanager.go

```diff
@@ -0,0 +1,137 @@
+// policyPrefix is prefixed to username and should be "edgex-user-" or policyPrefix
+// policy is a map that will be serialized as a policy attached to the identity
+func (m *UserManager) CreatePasswordUserWithPolicy(username string, password string, policyPrefix string, policy map[string]interface{}) error {
+
+	m.logger.Infof("creating policy, identity, and userpass binding for %s", username)
+
+	// Derive the name of the attached policy and create it
+
+	policyName := policyPrefix + username
+
+	policyBytes, err := json.Marshal(policy)
+	if err != nil {
+		m.logger.Errorf("failed encode policy for %s: %s", username, err.Error())
+		return err
+	}
+
+	if err := m.secretStoreClient.InstallPolicy(m.privilegedToken, policyName, string(policyBytes)); err != nil {
+		m.logger.Errorf("failed to install policy %s: %s", policyName, err.Error())
+		return err
+	}
+
+	// Create or update underlying vault identity
+	identityMetadata := map[string]string{
+		// we will also put a name claim in any generated JWT's
+		"name": username,
+	}
+	identityPolicies := []string{policyName}
+	identityId, err := m.secretStoreClient.CreateOrUpdateIdentity(m.privilegedToken, username, identityMetadata, identityPolicies)
+	if err != nil {
+		return err
+	}
+	if identityId == "" {
+		// Updating an entity doesn't return its ID (grr!), in that case, need to look it up
+		identityId, err = m.secretStoreClient.LookupIdentity(m.privilegedToken, username)
+		if err != nil {
+			return err
+		}
+	}
+
+	// When logging in "default" policy is added automatically to the token and identity_policy is inherited.
+	err = m.secretStoreClient.CreateOrUpdateUser(m.privilegedToken, m.userPassMountPoint, username, password, m.tokenTTL, []string{})
+	if err != nil {
+		return err
+	}
+
+	authHandle, err := m.secretStoreClient.LookupAuthHandle(m.privilegedToken, m.userPassMountPoint)
+	if err != nil {
+		return err
+	}
+
+	err = m.secretStoreClient.BindUserToIdentity(m.privilegedToken, identityId, authHandle, username)
+	if err != nil {
+		return err
+	}
+
+	// See https://developer.hashicorm.com/vault/docs/secrets/identity/identity-token#token-contents-and-templates
+	// for including custom claims.  We will include a claim identifying the calling EdgeX service
+	// We will use the OIDC standard "name" claim.
+	customClaims := fmt.Sprintf(`{"name": "%s"}`, username)
+	err = m.secretStoreClient.CreateOrUpdateIdentityRole(m.privilegedToken, username, m.jwtKeyName, customClaims, m.jwtTTL)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (m *UserManager) DeletePasswordUser(username string) error {
+
+	identityId, err := m.secretStoreClient.LookupIdentity(m.privilegedToken, username)
+	if err != nil {
+		return err
+	}
+	if identityId != "" {
+		err = m.secretStoreClient.DeleteIdentity(m.privilegedToken, username)
+		if err != nil {
+			return err
+		}
+	}
+
+	err = m.secretStoreClient.DeleteUser(m.privilegedToken, m.userPassMountPoint, username)
+	if err != nil {
+		return err
+	}
```

Comment:
should we make this transactional or atomic? if deleting user somehow failed then you lost identity for that user isn't it?

Reply:
I think its ok.  The ACL is attached to the identity so if you delete the identity the user is really gone.  You could attempt to log in as the user, but it would error out as it isn't bound to anything.  Moreover, if you recreated the user it would overwrite the identity and/or login, so a partial deletion would take up space but that's about it.

Reply:
ok, good to know that creating user would get a new identity.